### ISSUE TITLE
Creusot errors with "unsupported type" on functions returning impl FnMut

### ISSUE THREAD

**Issue by jhaye** [xldenis/creusot#539]
The project I'm attempting to use Creusot for uses [nom](https://crates.io/crates/nom) for parsing. The crate has many functions which have the return type `impl FnMut(I) -> IResult<I, O, E>`. An example of this is [`all_consuming(…)`](https://docs.rs/nom/7.1.1/nom/combinator/fn.all_consuming.html), .

Creusot fails with the following error:
```
error[creusot]: unsupported type impl FnMut(I)-> std::result::Result<(I, O), nom::Err<E>>
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/nom-7.1.1/src/combinator/mod.rs:382:1
    |
382 | / pub fn all_consuming<I, O, E: ParseError<I>, F>(mut f: F) -> impl FnMut(I) -> IResult<I, O, E>
383 | | where
384 | |   I: InputLength,
385 | |   F: Parser<I, O, E>,
    | |_____________________^

```

**Comment by xldenis** [comment]
Indeed, I haven't thought about how to integrate `impl Trait` into Creusot, though I don't think it is necessarily so difficult. What are you expecting to prove? I imagine that working with `nom` in proofs will be quite challenging since it's a large complex library without specifications.

**Comment by jhaye** [comment]
I'm trying to get it to compile with Creusot first. This is a side project of mine which I want to use to gauge Creusot's ability to integrate with an existing code base. It's a quasi copy of the Berkley parser.
If this is not supported yet, I can just discard the input parsing of the program and see if I can prove that it doesn't crash without them and then move on from there.

**Comment by xldenis** [comment]
You can mark functions as `#[trusted]` to tell creusot to skip them. This will still translate the signatures but should be used for parts of the code that can't / shouldn't be verified (like a parser)

**Comment by jhjourdan** [comment]
Waiting for an intern to work on this. This might be an interesting project.

**Comment by jhjourdan** [comment]
@xldenis : Am I right that #1351 fixes this ?

**Comment by xldenis** [comment]
Needs a test but yes.

**Comment by jhjourdan** [comment]
So I did some tests, and it seems thtat this works, except that we can no longer write resquires/ensures clauses...

See #1512.

I am closing this one, because the specific bug here is "fixed".
